tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -877,56 +879,111 @@ static void setLocationInfoForClangNode(ClangNode ClangNode,
-        auto OriginalStartOffset = SM.getLocOffsetInBuffer(
-            GeneratedSourceInfo->originalSourceRange.getStart(), OriginalBufID);
-        auto GeneratedStartOffset = SM.getLocOffsetInBuffer(
-            GeneratedSourceInfo->generatedSourceRange.getStart(), DeclBufID);
-        Location.Offset += OriginalStartOffset - GeneratedStartOffset;
-        std::tie(Location.Line, Location.Column) =
-            SM.getPresumedLineAndColumnForLoc(DeclLoc, DeclBufID);
+      if (auto FDR = FD->getParameterListSourceRange())
+        range = FDR;
+    }
+
+    unsigned bufID = SM.findBufferContainingLoc(range.Start);
+
+    // If this range is from a generated buffer, recursively "unexpand" macros
+    // to chase after where the macro was originally expanded from.
+    //
+    // However, we don't care about certain kinds of generated buffers, so save
+    // the original range and buffer ID so we can set location according to VD.
+    auto VDRange = range;
+    auto VDBufID = bufID;
+    while (auto *info = SM.getGeneratedSourceInfo(bufID)) {
+      switch (info->kind) {
+#define MACRO_ROLE(Name, Description)                                          \
+  case GeneratedSourceInfo::Name##MacroExpansion:
+#include "swift/Basic/MacroRoles.def"
+        if (auto *customAttr = info->attachedMacroCustomAttr)
+          range = customAttr->getRange();
+        else
+          range = ASTNode::getFromOpaqueValue(info->astNode).getSourceRange();
+        bufID = SM.findBufferContainingLoc(range.Start);
+        continue; // Continue while-loop to recursively un-expand macros
+
+      case GeneratedSourceInfo::ReplacedFunctionBody:
+        if (bufID == VDBufID) {
+          // The location was in a temporary source buffer that just contains
+          // the function body, which we created while reusing the ASTContext
+          // for the rest of the file. Set the location so that it maps back to
+          // the original file.
+          setLocationInfoForRange(SM, VDRange, VDBufID, Location,
+                                  /*presumed=*/true);
+          // Adjust offset according to generated buffer.
+          auto originalLoc = info->originalSourceRange.getStart();
+          auto originalBufID = SM.findBufferContainingLoc(originalLoc);
+          auto generatedLoc = info->generatedSourceRange.getStart();
+          auto generatedBufID = SM.findBufferContainingLoc(generatedLoc);
+          Location.Offset +=
+              SM.getLocOffsetInBuffer(originalLoc, originalBufID) -
+              SM.getLocOffsetInBuffer(generatedLoc, generatedBufID);
+        } else {
+          // We somehow encountered a replaced function body while looking
+          // through other macro expansions. Fall back to setting location based
+          // on VD's original source range because mapping location back to the
+          // original file might be tricky.
+          setLocationInfoForRange(SM, VDRange, VDBufID, Location);
```

Comment:
What did we do here previously 🤔?

Reply:
The previous behavior is buried in `getOriginalRange()`, which is to just bail out and use the original `SourceRange`:

https://github.com/swiftlang/swift/blob/00aa22f7b26abeaef60b37fc456d1c5d4f7c336b/lib/AST/Module.cpp#L879-L886

Now that you mention this, I'm not sure if that is the correct behavior, but I wasn't sure whether this scenario is even reachable (`ReplacedFunctionBody` would mostly be for source edits, right? So they shouldn't be embedded within other generated macro buffers?)

Reply:
Yeah, they shouldn't ever be embedded. TBH I would prefer checking for `ReplacedFunctionBody` specifically and re-using `getUnexpanded...` for the macro portion (which would also eliminate the duplicate code there).

Reply:
Ah... :( Okay